crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -584,39 +592,61 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
             let AccountBeforeTx { info: old_info, address } = account_before;
-            match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((old_info, new_info, HashMap::new()));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    // overwrite old account state.
-                    entry.get_mut().0 = old_info;
+
+            // If the address is not already in the bundle builder's state, add it.
+            if !bundle_builder.get_states().contains(&address) {
+                // Seek the address in the plain account state.
+                if let Some(new_info) = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1) {
+                    // Add the new info to the bundle builder.
+                    bundle_builder =
+                        bundle_builder.state_present_account_info(address, new_info.into());
                 }
             }
-            // insert old info into reverts.
-            reverts.entry(block_number).or_default().entry(address).or_default().0 = Some(old_info);
+
+            // Add the old account info to the bundle builder if it exists.
+            if let Some(old_info) = old_info {
+                bundle_builder =
+                    bundle_builder.state_original_account_info(address, old_info.into());
+            }
+
+            // Insert the old account info into the reverts.
+            bundle_builder = bundle_builder.revert_account_info(
+                block_number,
+                address,
+                Some(old_info.map(Into::into)),
+            );
         }
 
+        let mut state_storage: HashMap<Address, HashMap<U256, (U256, U256)>> = HashMap::new();
+
         // add storage changeset changes
         for (block_and_address, old_storage) in storage_changeset.into_iter().rev() {
             let BlockNumberAddress((block_number, address)) = block_and_address;
-            // get account state or insert from plain state.
-            let account_state = match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let present_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((present_info, present_info, HashMap::new()))
+            // If the address is not already in the bundle builder's state, add it.
+            if !bundle_builder.get_states().contains(&address) {
+                if let Some(present_info) =
+                    plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1)
+                {
+                    bundle_builder = bundle_builder
+                        .state_original_account_info(address, present_info.into())
+                        .state_present_account_info(address, present_info.into());
```

Comment:
looks like we setting functions that take &mut self would be useful on bundlebuilder

Reply:
Done here https://github.com/bluealloy/revm/pull/1527